### ISSUE TITLE
LambdaLR has incorrect multiplicative behavior when using torch.tensor LR

### ISSUE THREAD

**Issue by ad8e** [pytorch/pytorch#126854]
### 🐛 Describe the bug

```
import torch.nn as nn
import torch.optim as optim
import torch.optim.lr_scheduler as lr_scheduler
import torch

model = nn.Linear(10, 1)

optimizer = optim.Adam(model.parameters(), lr=torch.ones(1))

lambda1 = lambda epoch: 0.5
lr_scheduler = lr_scheduler.LambdaLR(optimizer, lambda1)

epochs = 10
for epoch in range(epochs):
    optimizer.step()
    lr_scheduler.step()
    print(optimizer.state_dict()["param_groups"][0]["lr"])
```

Output:
```
tensor([0.2500])
tensor([0.1250])
tensor([0.0625])
tensor([0.0312])
tensor([0.0156])
tensor([0.0078])
tensor([0.0039])
tensor([0.0020])
tensor([0.0010])
tensor([0.0005])
```
They should all be 0.5.

### Versions

```
PyTorch version: 2.4.0a0+ad67553
```
Does not appear on 2.3 or when using float LR.

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @vincentqb @jbschlosser @albanD @janeyx99 @crcrpar

**Comment by GdoongMathew** [comment]
I think this was introduced by #123753, https://github.com/pytorch/pytorch/blob/ba3b05fdf37ddbc3c301294d6a560a816335e717/torch/optim/lr_scheduler.py#L221

I haven't fully tested it yet, but I think it was because of using inplace operation, which also tempered its `initial_lr`

https://github.com/pytorch/pytorch/blob/ba3b05fdf37ddbc3c301294d6a560a816335e717/torch/optim/lr_scheduler.py#L78-L90

**Comment by GdoongMathew** [comment]
One quick fix I could think of, is to use deepcopy at https://github.com/pytorch/pytorch/blob/ba3b05fdf37ddbc3c301294d6a560a816335e717/torch/optim/lr_scheduler.py#L79-L80

```python
for group in optimizer.param_groups:
    group.setdefault("initial_lr", copy.deepcopy(group["lr"]))
```

**Comment by GdoongMathew** [comment]
BTW, I think this also affects every scheduler which use `_get_closed_form_lr`.
Will open a PR once finish testing.

**Comment by GdoongMathew** [comment]
@janeyx99 to answer your concerns, if accepting tensor is publicly acknowledged, I'd say that optimizer could also have this "side-effect"?
https://github.com/pytorch/pytorch/blob/ba3b05fdf37ddbc3c301294d6a560a816335e717/torch/optim/optimizer.py#L1022-L1028

**Comment by GdoongMathew** [comment]
I think there're still a few problems require discussion.
1. should this `deepcopy` operation kept in `LRScheduler`? or should it move to where Optimizer set their default value on their param_group?
2. should every hyperparameters accepts `torch.Tensor` as their input type in both `optimizer` and `lr_scheduler` module? if so, then probably every hyperparameters (like `betas` in `Adam`) should also deepcopying their input parameter as well?

**Comment by janeyx99** [comment]
1. Would moving it to the Optimizer even guarantee a fix? If the LRScheduler is allowed to read the lr of the Optimizer, it could technically read the deepcopy'd version and modify that anyway, right?
2. No, we haven't publicly made this an API for anything other than LR as we don't expect them to change, but this would not be against the direction we move towards. So right now this is not be a problem, but it'd be nice to have a solution that is well designed for both.

Separately, we should not use deepcopy on Tensors, but clone().detach() instead.

**Comment by GdoongMathew** [comment]
1. No, simply moving deepcopy to Optimizer wouldn't thoroughly fix this issue. I think both Optimizer and LRScheduler would need to have their own copy of LR to prevent from tensor being changed inplace from the outside.
2. Understood.

Thanks for the suggestion, its being added in the PR as well.

**Comment by GdoongMathew** [comment]
Hi @janeyx99 , shall we keep this issue open until a new PR for keeping this thread alive and visible to others?

> @GdoongMathew I've discussed the long term strategy more offline with @albanD to make sure we're heading the right direction, and I'd like to share our proposal for the desired end state. (This does not mean you need to rewrite the quick fix PR to implement this north star, but if you're interested in implementing it, I'm happy to review.)
> 
> Ultimately:
> 
> 1. we want to maintain the same semantics as when the lr was a Python number and not a Tensor:
>    a. no inplacing -> we should not have any `fill_` in our LRScheduler math/compute. Instead we should assign with `group["lr"] = ...`.
>    b. this goes hand-in-hand with cloning a copy during computation to assign back to the optimizer to avoid ever tangling the values.
> 2. we want to ensure that the LRScheduler preserves the "Tensorness" of the optimizer's hyperparams. In other words, if the Optimizer has a Tensor LR, the LRScheduler should always be giving it a Tensor LR to work with.
> 
> These invariants should give us a better way to reason through desired behavior of the LRScheduler.

**Comment by ad8e** [comment]
Another option is to make a new issue for tracking the general torch.tensor+optimizer design goals, since the specific problem in the title of this issue is fixed. https://github.com/pytorch/pytorch/issues/127699 is related. I'm fiddling with other optimizers (like laprop or EDM2 or schedule-free), which are likely to supplant AdamW, so optimizers and schedulers have some broad interfaces to be considered.

**Comment by janeyx99** [comment]
Making a new issue sounds appropriate as @ad8e suggested! And @ad8e I'm curious to hear about your experiments and consider what we could add to PyTorch optimizers next (through other issues).
